Crash on anonymous type parameters of implicit classes

```scala
object M {
  implicit class C[_](x: Int) {
    def m: Int = ???
  }
}
```

```
error: compiler crash: None.get
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:347)
	at scala.None$.get(Option.scala:345)
	at rsc.outline.Synthesizer$$anonfun$14.rsc$outline$Synthesizer$$anonfun$$synthesize$1(Synthesizer.scala:157)
	at rsc.outline.Synthesizer$$anonfun$14$$anonfun$15.apply(Synthesizer.scala:158)
	at rsc.outline.Synthesizer$$anonfun$14$$anonfun$15.apply(Synthesizer.scala:158)
	at scala.collection.immutable.List.map(List.scala:284)
	at rsc.outline.Synthesizer$$anonfun$14.rsc$outline$Synthesizer$$anonfun$$synthesize$1(Synthesizer.scala:158)
	at rsc.outline.Synthesizer$$anonfun$14.apply(Synthesizer.scala:164)
	at rsc.outline.Synthesizer$$anonfun$14.apply(Synthesizer.scala:154)
	at scala.collection.immutable.List.map(List.scala:284)
	at rsc.outline.Synthesizer.implicitClassConversion(Synthesizer.scala:154)
	at rsc.outline.Scheduler$$anonfun$stats$2.apply(Scheduler.scala:515)
	at rsc.outline.Scheduler$$anonfun$stats$2.apply(Scheduler.scala:503)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at rsc.outline.Scheduler.stats(Scheduler.scala:503)
	at rsc.outline.Scheduler.defnTemplate(Scheduler.scala:341)
	at rsc.outline.Scheduler.apply(Scheduler.scala:35)
	at rsc.outline.Scheduler$$anonfun$stats$2.apply(Scheduler.scala:512)
	at rsc.outline.Scheduler$$anonfun$stats$2.apply(Scheduler.scala:503)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at rsc.outline.Scheduler.stats(Scheduler.scala:503)
	at rsc.outline.Scheduler.source(Scheduler.scala:470)
	at rsc.outline.Scheduler.apply(Scheduler.scala:38)
	at rsc.Compiler$$anonfun$rsc$Compiler$$schedule$1.apply(Compiler.scala:108)
	at rsc.Compiler$$anonfun$rsc$Compiler$$schedule$1.apply(Compiler.scala:106)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at rsc.Compiler.rsc$Compiler$$schedule(Compiler.scala:106)
	at rsc.Compiler$$anonfun$phases$3.apply$mcV$sp(Compiler.scala:70)
	at rsc.Compiler$$anonfun$run$2.apply(Compiler.scala:36)
	at rsc.Compiler$$anonfun$run$2.apply(Compiler.scala:33)
	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:733)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
	at rsc.Compiler.run(Compiler.scala:33)
	at rsc.cli.Main$.process(Main.scala:35)
	at rsc.cli.Main$.main(Main.scala:14)
	at rsc.cli.Main.main(Main.scala)
```

I looked into this a bit to get a better idea of how `rsc` works.

For one, if we remove the `implicit` keyword, the compiler does not crash. However, dumping the generated outline class file with `scalap`, the `_` is converted to `anon$1`. This causes the following code to crash `rsc`, while it compiles with `scalac`:
```
object M {
  class C[_, anon$1](x: Int) {
    def m: Int = ???
  }
}
```
(Using `anon$2` instead of `anon$1` above does not cause `rsc` to crash.)

Another interesting thing is the following code compiles with `rsc` (with the two `_`'s being replaced by `anon$1` and `anon$2`), while `scalac` 2.11.11 returns an error:
```
object M {
  class C[_, _](x: Int) {
    def m: Int = ???
  }
}
```
```
$ scalac M.scala
M.scala:2: error: _ is already defined as type _
    class C[_, _](x: Int) {
               ^
one error found
```

It seems that a direct `_` type parameter in a class or method definition is handled almost exactly like a type variable, except that no binding is introduced for that type variable within the scope of the class. Perhaps we need to adjust the parser to differentiate between this case and higher kinded type parameters (like `class C[Map[_, _]] { ... }`).

----

Now, the reason that the `implicit class` is handled differently from a plain `class` is there's a `match`-`case` in `Scheduler.stats()` that specifically looks for an `implicit class`, and when found, calls `synthesizer.implicitClassConversion(currEnv, tree)`. The crash this ticket describes is at some point within this call.
```
        tree match {
          case tree: DefnClass if tree.hasImplicit =>
            synthesizer.implicitClassConversion(currEnv, tree)
          case _: DefnCtor | _: PrimaryCtor =>
            ()
          case tree: DefnDef if tree.hasDefaultParams =>
            synthesizer.defaultGetters(treeEnv, tree)
          case _ =>
            ()
        }
```

So the problem is at least partly in `rsc.outline.Synthesizer`. I noticed a `FIXME` comment at the top of that class that points to #98, so overall it seems the `Synthesizer` class needs some improvement.

--- a/examples/semantic/src/main/scala/366.scala
+++ b/examples/semantic/src/main/scala/366.scala
@@ -1,0 +1,7 @@
+package ticket366
+
+object M {
+  implicit class C[_](x: Int) {
+    def m: Int = ???
+  }
+}

--- a/rsc/src/main/scala/rsc/outline/Scheduler.scala
+++ b/rsc/src/main/scala/rsc/outline/Scheduler.scala
@@ -666,6 +666,8 @@ final class Scheduler private (
         apply(treeEnv, tree)
         tree match {
           case tree: DefnClass if tree.hasImplicit =>
+            // When we encounter an implicit class, we must synthesize an implicit factory method
+            // of the same name located in the same scope.
             synthesizer.implicitClassConversion(currEnv, tree)
           case _: DefnCtor | _: PrimaryCtor =>
             ()

--- a/rsc/src/main/scala/rsc/outline/Synthesizer.scala
+++ b/rsc/src/main/scala/rsc/outline/Synthesizer.scala
@@ -148,6 +148,15 @@ final class Synthesizer private (
     enumValues(env, tree)
   }
 
+  /**
+   * Given an implicit class, synthesize an implicit factory method for it in
+   * the same scope. Note that if the implicit class has an anonymous type
+   * parameter (e.g., `implicit class C[_] { }`), the `_` will be replaced by
+   * `Any` in the factory method's return type (e.g., `C[Any]`).
+   *
+   * @param env
+   * @param tree
+   */
   def implicitClassConversion(env: Env, tree: DefnClass): Unit = {
     val mods = tree.mods.filter(_.isInstanceOf[ModAccess]) :+ ModImplicit()
     val id = TermId(tree.id.value)
@@ -175,8 +184,19 @@ final class Synthesizer private (
     }
     val ret = {
       val core = tree.id
-      if (tparams.isEmpty) Some(core)
-      else Some(TptParameterize(core, tparams.map(_.id.asInstanceOf[TptId])))
+      // Convert the implicit class's type parameters to the factory method return type parameters.
+      // This is necessary because the type parameters of a class or method definition can include
+      // constraints like upper and lower bounds, while a method return type can only include
+      // concrete type names and type variables that were bound in the method's type parameters.
+      val retTparams = tparams.map(_.id match {
+        case tptId: TptId => tptId
+        // AnonId represents `_`, which doesn't make sense as a type parameter in the return type;
+        // replace it by `Any`, i.e., `_root_.scala.Any`
+        case anonId: AnonId => TptId("Any").withSym(AnyClass)
+        case nonTptNamedId: NamedId => crash(nonTptNamedId.pos, nonTptNamedId)
+      })
+      if (retTparams.isEmpty) Some(core)
+      else Some(TptParameterize(core, retTparams))
     }
     val rhs = Some(TermStub())
     val meth = DefnMethod(mods, id, tparams, paramss, ret, rhs)

